File: content/snippets/scala3/problems_top_k_frequent.scala
object TopKFrequent:
  def topKFrequent(nums: Vector[Int], k: Int): Vector[Int] =
    nums
      .groupBy(identity)
      .view
      .mapValues(_.length)
      .toVector
      .sortBy { case (_, count) => -count }
      .take(k)
      .map { case (value, _) => value }

  @main def main(): Unit =
    val nums = Vector(1, 1, 1, 2, 2, 3, 4, 4, 4, 4, 5, 5)
    val k = 2
    val result = topKFrequent(nums, k)

    println("top k frequent: " + result.mkString(" "))

File: content/snippets/cpp/problems_top_k_frequent.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <unordered_map>
#include <utility>
#include <vector>

std::vector<int> topKFrequent(const std::vector<int>& nums, int k) {
    std::unordered_map<int, int> freq;
    for (int n : nums) {
        ++freq[n];
    }

    using Entry = std::pair<int, int>;  // (frequency, value)
    auto cmp = [](const Entry& a, const Entry& b) { return a.first > b.first; };
    std::priority_queue<Entry, std::vector<Entry>, decltype(cmp)> minHeap(cmp);

    for (const auto& [value, count] : freq) {
        minHeap.push({count, value});
        if (static_cast<int>(minHeap.size()) > k) {
            minHeap.pop();
        }
    }

    std::vector<int> result;
    while (!minHeap.empty()) {
        result.push_back(minHeap.top().second);
        minHeap.pop();
    }
    std::reverse(result.begin(), result.end());
    return result;
}

int main() {
    const std::vector<int> nums = {1, 1, 1, 2, 2, 3, 4, 4, 4, 4, 5, 5};
    const int k = 2;

    const auto result = topKFrequent(nums, k);
    std::cout << "top k frequent:";
    for (int n : result) {
        std::cout << ' ' << n;
    }
    std::cout << '\n';
    return 0;
}
